src/geometry/rtree.cpp

```diff
@@ -0,0 +1,290 @@
+}
+template <typename coord_system>
+constexpr auto RTree<coord_system>::get_id(doc_type const& doc) -> t_docId {
+  return doc.second;
+}
+
+template <typename coord_system>
+auto RTree<coord_system>::lookup(t_docId id) const
+    -> std::optional<std::reference_wrapper<const geom_type>> {
+  if (auto it = docLookup_.find(id); it != docLookup_.end()) {
+    return it->second;
+  }
+  return std::nullopt;
+}
+template <typename coord_system>
+auto RTree<coord_system>::lookup(doc_type const& doc) const
+    -> std::optional<std::reference_wrapper<const geom_type>> {
+  return lookup(get_id(doc));
+}
+
+template <typename coord_system>
+auto RTree<coord_system>::from_wkt(std::string_view wkt) const -> geom_type {
+  geom_type geom{};
+  if (wkt.starts_with("POI")) {
+    geom = bg::from_wkt<point_type>(wkt.data());
+  } else if (wkt.starts_with("POL")) {
+    geom = bg::from_wkt<poly_type>(wkt.data());
+  } else {
+    throw std::runtime_error{"unknown geometry type"};
+  }
+  std::visit(
+      [](auto&& geom) {
+        if (bg::is_empty(geom)) {
+          throw std::runtime_error{"attempting to create empty geometry"};
+        }
+        if (!bg::is_valid(geom)) {  // TODO: GEOMETRY - add flag to allow user to ascertain
+                                    // validity of input
+          throw std::runtime_error{"invalid geometry"};
+        }
+      },
+      geom);
+  return geom;
+}
+
+template <typename coord_system>
+constexpr auto geometry_reporter = [](auto&& geom) {
+  using point_type = typename RTree<coord_system>::point_type;
+  if constexpr (std::is_same_v<point_type, std::decay_t<decltype(geom)>>) {
+    return 0ul;
+  } else {
+    auto const& inners = geom.inners();
+    auto outer_size = geom.outer().get_allocator().report();
+    return std::accumulate(
+        inners.begin(), inners.end(), outer_size,
+        [](std::size_t acc, auto&& hole) { return acc + hole.get_allocator().report(); });
+  }
+};
+
+template <typename coord_system>
+void RTree<coord_system>::insert(geom_type const& geom, t_docId id) {
+  rtree_.insert(make_doc(geom, id));
+  docLookup_.insert(LUT_value_type{id, geom});
+  alloc_.allocated_ += std::visit(geometry_reporter<coord_system>, geom);
+}
+
+template <typename coord_system>
+int RTree<coord_system>::insertWKT(const char* wkt, std::size_t len, t_docId id,
+                                   RedisModuleString** err_msg) {
+  try {
+    auto geom = from_wkt(std::string_view{wkt, len});
+    insert(geom, id);
+    return 0;
+  } catch (const std::exception& e) {
+    if (err_msg) {
+      *err_msg = RedisModule_CreateString(nullptr, e.what(), strlen(e.what()));
+    }
+    return 1;
+  }
+}
+
+template <typename coord_system>
+bool RTree<coord_system>::remove(t_docId id) {
+  if (auto geom = lookup(id); geom.has_value()) {
+    rtree_.remove(make_doc(geom.value(), id));
+    alloc_.allocated_ -= std::visit(geometry_reporter<coord_system>, geom.value().get());
+    docLookup_.erase(id);
```

Comment:
The lookup is using the same allocator, so it would be tracked by it? (no need to decrease the count manually)

Reply:
same deal as above